src/store/singing.ts

```diff
@@ -946,241 +939,378 @@ export const singingStore = createPartialStore<SingingStoreTypes>({
+            // 歌い方と音声を削除する
+            if (phrase.singingStyleKey != undefined) {
+              commit("DELETE_SINGING_STYLE", {
+                singingStyleKey: phrase.singingStyleKey,
+              });
+            }
+            if (phrase.singingVoiceKey != undefined) {
+              singingVoices.delete(phrase.singingVoiceKey);
+            }
+
+            // 音源とシーケンスの接続を解除して削除する
+            const sequence = sequences.get(phraseKey);
+            if (sequence) {
+              if (sequence.type === "note") {
+                sequence.instrument.output.disconnect();
+              } else if (sequence.type === "audio") {
+                sequence.audioPlayer.output.disconnect();
+              } else {
+                throw new Error("Unknown type of sequence.");
+              }
+              transportRef.removeSequence(sequence);
+              sequences.delete(phraseKey);
+            }
           }
         }
-        for (const key of state.phrases.keys()) {
-          if (!foundPhrases.has(key)) {
-            commit("DELETE_PHRASE", { phraseKey: key });
-
-            // フレーズ削除時の処理
-            const phraseData = phraseDataMap.get(key);
-            if (!phraseData) {
-              throw new Error("phraseData is undefined");
+
+        const phrases = new Map<string, Phrase>();
+
+        for (const [notesHash, foundPhrase] of foundPhrases) {
+          const phraseKey = notesHash;
+          const existingPhrase = state.phrases.get(phraseKey);
+          if (!existingPhrase) {
+            // 新しいフレーズの場合
+            phrases.set(phraseKey, foundPhrase);
+            continue;
+          }
+
+          // すでに存在するフレーズの場合
+          // 再レンダリングする必要があるかどうかをチェックする
+          // シンガーが未設定の場合、とりあえず常に再レンダリングする
+          // 音声合成を行う必要がある場合、現在フレーズに設定されている音声を削除する
+          // 歌い方の推論も行う必要がある場合、現在フレーズに設定されている歌い方を削除する
+          // TODO: リファクタリングする
+          const phrase = { ...existingPhrase };
+          if (!singer || phrase.state === "COULD_NOT_RENDER") {
+            if (phrase.singingStyleKey != undefined) {
+              commit("DELETE_SINGING_STYLE", {
+                singingStyleKey: phrase.singingStyleKey,
+              });
+              phrase.singingStyleKey = undefined;
             }
-            if (phraseData.source) {
-              phraseData.source.output.disconnect();
+          } else if (phrase.singingStyleKey != undefined) {
+            const calculatedSingingStyleSourceHash =
+              await calculateSingingStyleSourceHash({
+                engineId: singer.engineId,
+                tpqn,
+                tempos,
+                notes: phrase.notes,
+                keyRangeAdjustment,
+                volumeRangeAdjustment,
+                frameRate: frameRates[singer.engineId],
+                restDurationSeconds,
+              });
+            const singingStyleSourceHash = phrase.singingStyleKey;
+            if (singingStyleSourceHash !== calculatedSingingStyleSourceHash) {
```

Comment:
ここあたりのコードは必ず`singingStyleSource`系の話なので、単純にoldHashやnewHashと書いちゃえる気もしました！
```suggestion
            const oldHash = phrase.singingStyleKey;
            if (oldHash !== newHash) {
```

Reply:
`hash`と`calculatedHash`にしました！